### ISSUE TITLE
Feature proposal: More concise layout syntax

### ISSUE THREAD

**Issue by janfreyberg** [jupyter-widgets/ipywidgets#2895]
At the moment, "laying out" a few widgets to create a GUI can result in a lot of code: If I want, e.g., a GUI that has two columns, and in each column, a variable number of rows which might have columns too, I quickly end up with many nested `HBox` and `VBox` widgets.

Since this seems like a common use case to me, I wondered if it would be possible to implement a more concise method of laying out widgets, similar to how the [altair library does it](https://altair-viz.github.io/user_guide/compound_charts.html#horizontal-concatenation)?

For example, like this:

```python
import ipywidgets as widgets

def hconcat(self, other):
    if isinstance(self, widgets.HBox):
        self.children = self.children + (other,)
        return self
    elif isinstance(other, widgets.HBox):
        other.children = (self,) + other.children
        return other
    else:
        return widgets.HBox([self, other])


def vconcat(self, other):
    if isinstance(self, widgets.VBox):
        self.children = self.children + (other,)
        return self
    elif isinstance(other, widgets.VBox):
        other.children = (self,) + other.children
        return other
    else:
        return widgets.VBox([self, other])

widgets.Widget.__or__ = hconcat
widgets.Widget.__and__ = vconcat

(
    (widgets.Button(description='left') | widgets.Button(description='right'))
    & widgets.Button(description='middle')
    & (widgets.Button(description='bottom left') | widgets.Button(description='bottom right'))
)
```

The resulting widget looks like:
```
VBox(
    children=(
        HBox(
            children=(
                Button(description="left", style=ButtonStyle()),
                Button(description="right", style=ButtonStyle()),
            )
        ),
        Button(description="middle", style=ButtonStyle()),
        HBox(
            children=(
                Button(description="bottom left", style=ButtonStyle()),
                Button(description="bottom right", style=ButtonStyle()),
            )
        ),
    )
)
```

I'm happy to implement this but wanted to raise an issue first to see if it would be rejected for being counterintuitive (which I understand).

**Comment by janfreyberg** [comment]
I agree that it is more explicit. And, while the `|` operator is relatively obvious, the `&` operator is not.

I think for me in particular, as I use the black code formatter, the nested `Box` classes end up taking up a lot of lines. But that is of course my choice.

For reference, similar concepts exist in the R universe, where they use `/` for vertical stacking:
https://patchwork.data-imaginist.com/

**Comment by maartenbreddels** [comment]
> For reference, similar concepts exist in the R universe, where they use `/` for vertical stacking:
> https://patchwork.data-imaginist.com/

Big fan of using `|` and `/`.

What I had in mind, was to use some kind of context manager, and that the widget calls whatever object is set to do the layout, so we can have different combinations.
e.g. (hypothetical code)
```python
from ipyvuetify.patchwork import responsive_layout

with responsive_layout(md=6, lg=4, ... whatever parameters):
  widget = (Button() | v.Btn()) / bqplot.Figure()
display(widget)
```

This would mean core ipywidgets would need to support it, (forwarding operator calls), but can outsource it to different implementations.

**Comment by janfreyberg** [comment]
That would be really cool, but I would probably expect there to be a "default" implementation in ipywidgets so that you don't get errors if you don't have such a context manager.

Otherwise, a context manager like that could also just "monkeypatch" the operator overloading, similar to what I did in the first post. This wouldn't require any changes in `ipywidgets`.
